src/storage-types/src/sources/mysql.rs

```diff
@@ -159,3 +173,259 @@ impl RustType<ProtoMySqlSourceDetails> for MySqlSourceDetails {
+
+    fn summarize(_path: Self::Summary) -> <() as Timestamp>::Summary {}
+}
+
+/// This type is used to represent the the progress of each MySQL GTID 'source_id' in the
+/// ingestion dataflow.
+///
+/// A MySQL GTID consists of a source_id (UUID) and transaction_id (non-zero u64).
+///
+/// For the purposes of this documentation effort, the term "source" refers to a MySQL
+/// server that is being replicated from: https://dev.mysql.com/doc/refman/8.0/en/replication-gtids-concepts.html
+///
+/// Each source_id represents a unique MySQL server, and the transaction_id
+/// monotonically increases for each source, representing the position of the transaction
+/// relative to other transactions on the same source.
+///
+/// Paritioining is by source_id which can be a singular UUID to represent a single source_id
+/// or a range of UUIDs to represent multiple sources.
+///
+/// The value of the partition is the NEXT transaction_id that we expect to see for the
+/// corresponding source(s), represented a GTIDState::Next(transaction_id).
+///
+/// GTIDState::Absent represents that no transactions have been seen for the
+/// corresponding source(s).
+///
+/// A complete Antichain of this type represents a frontier of the future transactions
+/// that we might see.
+pub type GtidPartition = Partitioned<Uuid, GTIDState>;
+
+impl SourceTimestamp for GtidPartition {
+    fn encode_row(&self) -> Row {
+        let ts = match self.timestamp() {
+            GTIDState::Absent => Datum::Null,
+            GTIDState::Active(id) => Datum::from(id.get()),
+        };
+        Row::pack(&[
+            Datum::from(self.interval().lower),
+            Datum::from(self.interval().upper),
+            ts,
+        ])
+    }
+
+    fn decode_row(row: &Row) -> Self {
+        let mut datums = row.iter();
+        match (datums.next(), datums.next(), datums.next(), datums.next()) {
+            (Some(Datum::Uuid(lower)), Some(Datum::Uuid(upper)), Some(Datum::UInt64(ts)), None) => {
+                match ts {
+                    0 => Partitioned::new_range(lower, upper, GTIDState::Absent),
+                    ts => Partitioned::new_range(
+                        lower,
+                        upper,
+                        GTIDState::Active(NonZeroU64::new(ts).unwrap()),
+                    ),
+                }
+            }
+            (Some(Datum::Uuid(lower)), Some(Datum::Uuid(upper)), Some(Datum::Null), None) => {
+                Partitioned::new_range(lower, upper, GTIDState::Absent)
+            }
+            _ => panic!("invalid row {row:?}"),
+        }
+    }
+}
+
+/// Parses a GTID Set string received from a MySQL server (e.g. from @@gtid_purged or @@gtid_executed).
+///
+/// Returns the frontier of all future GTIDs that are not contained in the provided GTID set.
+///
+/// TODO(roshan): Add compatibility for MySQL 8.3 'Tagged' GTIDs
+pub fn gtid_set_frontier(gtid_set_str: &str) -> Result<Antichain<GtidPartition>, io::Error> {
```

Comment:
We should write a bunch of unit tests for a variety of inputs to make sure the parsing logic works correctly